### PR TITLE
Dan/fix incorrect service grouping

### PR DESCRIPTION
**Pull request by danielsdeleo**
### :nut_and_bolt: Description

A customer reported that services are not assigned to the correct groups when they have different application settings. I was able to reproduce pretty easily by running:

```
applications_publish_supervisor_message --application ONE --sup-id 1111
applications_publish_supervisor_message --application TWO --sup-id 2222
applications_publish_supervisor_message --application THREE --sup-id 3333
```

On insert of a new service, the problem was that we were only looking at service groups based on their name, whereas we needed to be checking both the name and deployment id to get the correct row back.

After applying a fix for that problem, I investigated how the system would behave if the user has existing data that is incorrect due to this insertion bug. We recently added some code to modify the application and environment names if they change, but I've discovered that it was subtly incorrect: the existing code updates the deployment and/or service group values to match what was given in the latest health check message, but this changes the app name/environment/service group name for all services associated with the deployment/service group. I've therefore changed the logic so that when a health check message indicates that a service should belong to a different service group/deployment than what is currently in the database, we create a new deployment/service group, if necessary, then move the service over to the correct deployment and service group. I have confirmed that if there is incorrect data in the database resulting from the #1000 bug, it will be repaired as new health checks come in and the update logic gets applied.

### :athletic_shoe: Demo Script / Repro Steps

To repro bug #1000, run this in the studio with the packages from the dev channel:

```
applications_publish_supervisor_message --application ONE --sup-id 1111
applications_publish_supervisor_message --application TWO --sup-id 2222
applications_publish_supervisor_message --application THREE --sup-id 3333
```

In the UI you will see all 3 services in one row in the UI which indicates they are all in the same application.

To repro the fix, build applications service and wait for it to get deployed. Then re-run the `applications_publish_supervisor_message` step from above and you should see 3 service groups in the UI.

### :chains: Related Resources

#1000 

### :white_check_mark: Checklist

- [x] Necessary tests added/updated?
- [x] Code actually executed?
- [x] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/applications-service/integration_test/ingestion_test.go [hunk 1]
```diff
@@ -672,3 +673,60 @@ func TestIngestConcurrencySafe(t *testing.T) {
...
+	require.NoError(t, err)
+	suite.Ingester.IngestMessage(bytes3)
+	eventsProcessed++
+	suite.WaitForEventsToProcess(eventsProcessed)
+
+	// TODO: replace printf with test to assert we have 2 services w/ 2 supervisors in 2 service groups in 2 deployments
+	svcList, err = suite.StorageClient.GetServices("name", true, 1, 5, nil)
+	require.NoError(t, err)
+	for _, svc := range svcList {
+		fmt.Printf("%+v\n", svc)
```
Comment (danielsdeleo) [inline components/applications-service/integration_test/ingestion_test.go]: Prior to fixing the update logic, this prints (look at app and env):

```
&{ID:48 SupMemberID:1 Origin:core Name:db Version:0.1.0 Release:20200101121212 Status: Health:OK Group:db.default Fqdn: Application:newApp Environment:newEnv Channel: Site:us PreviousHealth:NONE UpdateStrategy:NONE LastEventOccurredAt:2019-07-24 20:36:38.118062 +0000 +0000 HealthUpdatedAt:2019-07-24 20:36:38.147417 +0000 +0000}
&{ID:47 SupMemberID:2 Origin:core Name:db Version:0.1.0 Release:20200101121212 Status: Health:OK Group:db.default Fqdn: Application:newApp Environment:newEnv Channel: Site:us PreviousHealth:NONE UpdateStrategy:NONE LastEventOccurredAt:2019-07-24 20:36:38.16734 +0000 +0000 HealthUpdatedAt:2019-07-24 20:36:38.14151 +0000 +0000}
```

After fixing, it prints:

```
&{ID:51 SupMemberID:1 Origin:core Name:db Version:0.1.0 Release:20200101121212 Status: Health:OK Group:db.default Fqdn: Application:app Environment:test-env Channel: Site:us PreviousHealth:NONE UpdateStrategy:NONE LastEventOccurredAt:2019-07-24 21:31:46.952817 +0000 +0000 HealthUpdatedAt:2019-07-24 21:31:46.986629 +0000 +0000}
&{ID:50 SupMemberID:2 Origin:core Name:db Version:0.1.0 Release:20200101121212 Status: Health:OK Group:db.default Fqdn: Application:newApp Environment:newEnv Channel: Site:us PreviousHealth:NONE UpdateStrategy:NONE LastEventOccurredAt:2019-07-24 21:31:46.997118 +0000 +0000 HealthUpdatedAt:2019-07-24 21:31:46.977513 +0000 +0000}
```

File: components/applications-service/pkg/storage/postgres/hab_event.go [hunk 2]
```diff
@@ -172,43 +172,135 @@ func (db *Postgres) IngestHealthCheckEventWithoutMetrics(event *habitat.HealthCh
...
+	// the one we found. This will correctly handle the case where a service is
+	// in the same service group and deployment as it is now, or the case that
+	// the service was "moved" to a different deployment and service group that
+	// already exist.
+	//
+	// If there is no existing deploymentServiceGroup that matches the metadata,
+	// then the service has been moved to a new deployment/service group. In that
+	// case, we need to create them.
+	// After that, we need to check if the old deployment and service group have
+	// no more related services, and delete them if so.
```
Comment (danielsdeleo) [inline components/applications-service/pkg/storage/postgres/hab_event.go]: The changes here are worth a careful read 🙏

File: components/applications-service/pkg/storage/postgres/hab_event.go [hunk 3]
```diff
@@ -172,43 +172,135 @@ func (db *Postgres) IngestHealthCheckEventWithoutMetrics(event *habitat.HealthCh
...
 
-	deploy, err := db.getDeployment(svc.DeploymentID)
-	if err != nil {
-		return errors.Wrap(err, "unable to update tables")
+	// When the service belongs to a deployment+service_group that exist (either
+	// the same one as before, or a different one):
+	if err == nil {
+		// The service could be in the same service_group and deployment as it was
+		// before, but it could also be in a new one.
+		// TODO/FIXME: updateService needs to adjust the deployment_id and group_id
```
Comment (danielsdeleo) [inline components/applications-service/pkg/storage/postgres/hab_event.go]: TODONE

File: components/applications-service/pkg/storage/postgres/hab_event.go [hunk 4]
```diff
@@ -395,7 +493,15 @@ func (db *Postgres) insertNewService(
 		}
 
 		// 2) Service Group
-		gid, exist := db.getServiceGroupID(svcMetadata.GetServiceGroup())
+		// FIXME that's your problem right there :P
+		// we need to find service group by name and deployment id (done)
+		// TODO after this:
+		// * regression test
+		// * see how it behaves when running with existing bad data
+		// * does updateService work? (seems to, but be more systematic)
```
Comment (danielsdeleo) [inline components/applications-service/pkg/storage/postgres/hab_event.go]: the answer is no, see above.

File: components/applications-service/pkg/storage/postgres/hab_event.go [hunk 5]
```diff
@@ -395,7 +493,15 @@ func (db *Postgres) insertNewService(
 		}
 
 		// 2) Service Group
-		gid, exist := db.getServiceGroupID(svcMetadata.GetServiceGroup())
+		// FIXME that's your problem right there :P
+		// we need to find service group by name and deployment id (done)
+		// TODO after this:
+		// * regression test
+		// * see how it behaves when running with existing bad data
```
Comment (danielsdeleo) [inline components/applications-service/pkg/storage/postgres/hab_event.go]: need to confirm, but I think fixing the update service logic will fix it

File: components/applications-service/pkg/storage/postgres/hab_event.go (hunk 3 above)
Comment (kmacgugan) [reply components/applications-service/pkg/storage/postgres/hab_event.go]: If it is TODONE can we remove the comment?

File: components/applications-service/integration_test/service_groups_sorting_test.go [hunk 6]
```diff
@@ -218,11 +218,13 @@ func TestGetServiceGroupsSortedPercent(t *testing.T) {
 	response, err := suite.ApplicationsServer.GetServiceGroups(ctx, request)
 	assert.Nil(t, err)
 
-	if assert.Equal(t, 4, len(response.ServiceGroups)) {
+	if assert.Equal(t, 6, len(response.ServiceGroups)) {
```
Comment (danielsdeleo) [inline components/applications-service/integration_test/service_groups_sorting_test.go]: Sorting behavior shouldn't be changed but these test were relying on the bug behavior so they had to be changed.

File: components/applications-service/pkg/storage/postgres/db.go [hunk 7]
```diff
@@ -111,13 +107,6 @@ type service struct {
 	HealthUpdatedAt     time.Time `db:"health_updated_at"`
 	CreatedAt           time.Time `db:"-"`
 	UpdatedAt           time.Time `db:"-"`
-
-	// (internal) use it to know if the service needs an update or not
-	needUpdate bool `db:"-"`
-}
-
-func (svc *service) NeedUpdate() bool {
-	return svc.needUpdate
```
Comment (danielsdeleo) [inline components/applications-service/pkg/storage/postgres/db.go]: We update the service every time so we can find when it disconnects, so this was just always returning true in practice.

File: components/applications-service/pkg/storage/postgres/db.go [hunk 8]
```diff
@@ -144,13 +133,6 @@ type serviceGroup struct {
 	DeploymentID int32     `db:"deployment_id"`
 	CreatedAt    time.Time `db:"-"`
 	UpdatedAt    time.Time `db:"-"`
-
-	// (internal) use it to know if the service-group needs an update or not
-	needUpdate bool `db:"-"`
-}
-
-func (sg *serviceGroup) NeedUpdate() bool {
-	return sg.needUpdate
```
Comment (danielsdeleo) [inline components/applications-service/pkg/storage/postgres/db.go]: Service group and deployment shouldn't be changed ever, we should just move services between them.

File: components/applications-service/pkg/storage/postgres/hab_event.go [hunk 9]
```diff
@@ -172,62 +171,134 @@ func (db *Postgres) IngestHealthCheckEventWithoutMetrics(event *habitat.HealthCh
...
+ WHERE sg.name = $1 AND d.app_name = $2 AND d.environment = $3;
+`
+
+	maybeCleanupServiceGroup = `
+DELETE FROM service_group WHERE service_group.id = $1 AND (NOT EXISTS (SELECT 1 FROM service WHERE service.group_id = service_group.id ))
+`
+	maybeCleanupDeployment = `
+DELETE FROM deployment WHERE deployment.id = $1 AND (NOT EXISTS (SELECT 1 FROM service WHERE service.deployment_id = deployment.id ))
+`
+)
```
Comment (danielsdeleo) [inline components/applications-service/pkg/storage/postgres/hab_event.go]: oops, need to move these things to not-here.

File: components/applications-service/pkg/storage/postgres/hab_event.go [hunk 10]
```diff
@@ -172,62 +171,134 @@ func (db *Postgres) IngestHealthCheckEventWithoutMetrics(event *habitat.HealthCh
...
+//
+// If we did find a matching deployment and service group, and it's not what is
+// currently set on the service, we update the foreign key relations to the
+// other deployment and service group when we apply updates for the service's
+// other fields.
+//
+// If we changed the service's deployment and service group, then we run a
+// cleanup query to ensure the prior deployment and service group are deleted
+// if they no longer have associated services.
+func (db *Postgres) updateServiceAndRelations(
```
Comment (danielsdeleo) [inline components/applications-service/pkg/storage/postgres/hab_event.go]: open to suggestions on how to break this function down into more readable parts

**Pull request merged**